File: backend/app/schemas/tendering.py
from __future__ import annotations

from datetime import date
from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import EmailStr
from pydantic import Field


TenderStatus = Literal["draft", "sent", "bids_in", "compared", "awarded", "cancelled"]
TenderDocumentType = Literal["plan", "markup", "spec"]
TenderBidStatus = Literal["invited", "received", "reviewed", "cancelled"]


class CategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    slug: str


class ContractorCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    contact_name: str | None = Field(default=None, max_length=255)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=50)
    address: str | None = None
    notes: str | None = None
    active: bool = True
    category_ids: list[UUID] = Field(default_factory=list)


class ContractorUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    contact_name: str | None = Field(default=None, max_length=255)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=50)
    address: str | None = None
    notes: str | None = None
    active: bool | None = None
    category_ids: list[UUID] | None = None


class ContractorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    contact_name: str | None = None
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    notes: str | None = None
    active: bool
    categories: list[CategoryOut]
    created_at: datetime
    updated_at: datetime


class TenderPackageCreate(BaseModel):
    category_id: UUID
    scope_description: str = Field(min_length=1)
    due_date: date | None = None


class TenderPackageUpdate(BaseModel):
    category_id: UUID | None = None
    scope_description: str | None = Field(default=None, min_length=1)
    status: TenderStatus | None = None
    due_date: date | None = None


class TenderDocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tender_package_id: UUID
    document_type: TenderDocumentType
    file_path: str
    original_filename: str
    uploaded_at: datetime


class TenderBidDocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tender_bid_id: UUID
    file_path: str
    original_filename: str
    uploaded_at: datetime


class TenderBidCreate(BaseModel):
    contractor_id: UUID


class TenderLineItem(BaseModel):
    description: str = Field(min_length=1)
    amount: Decimal


class TenderBidUpdate(BaseModel):
    quote_amount: Decimal | None = Field(default=None, ge=0)
    extracted_line_items: list[TenderLineItem] | None = None
    excluded_scope_notes: str | None = None
    reviewer_notes: str | None = None


class TenderBidOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tender_package_id: UUID
    contractor_id: UUID
    contractor: ContractorOut
    status: TenderBidStatus
    quote_amount: Decimal | None = None
    extracted_amount: Decimal | None = None
    extracted_line_items: list[TenderLineItem] | None = None
    excluded_scope_notes: str | None = None
    reviewer_notes: str | None = None
    invited_at: datetime | None = None
    received_at: datetime | None = None
    documents: list[TenderBidDocumentOut]
    created_at: datetime
    updated_at: datetime


class TenderAwardCreate(BaseModel):
    winning_bid_id: UUID
    budget_id: UUID
    budget_line_id: UUID
    award_instructions: str = Field(min_length=1)
    project_start_date: date
    contractor_start_date: date


class PurchaseOrderSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    po_number: str
    budget_id: UUID
    budget_line_id: UUID
    description: str
    amount: Decimal
    status: str


class TenderAwardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tender_package_id: UUID
    winning_bid_id: UUID
    po_id: UUID
    award_instructions: str
    project_start_date: date
    contractor_start_date: date
    awarded_at: datetime
    purchase_order: PurchaseOrderSummary | None = None


class TenderPackageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    property_id: UUID
    category_id: UUID
    category: CategoryOut
    scope_description: str
    status: TenderStatus
    due_date: date | None = None
    documents: list[TenderDocumentOut]
    bids: list[TenderBidOut] = Field(default_factory=list)
    award: TenderAwardOut | None = None
    created_at: datetime
    updated_at: datetime


class TenderComparisonOut(BaseModel):
    package: TenderPackageOut
    bids: list[TenderBidOut]
    bid_count: int
    received_count: int

File: frontend/src/types/tendering.ts
export interface ContractorCategory { id: string; name: string; slug: string }
export interface Contractor {
  id: string; name: string; contact_name?: string | null; email?: string | null; phone?: string | null;
  address?: string | null; notes?: string | null; active: boolean; categories: ContractorCategory[];
  created_at: string; updated_at: string;
}
export interface ContractorPayload {
  name: string; contact_name?: string | null; email?: string | null; phone?: string | null;
  address?: string | null; notes?: string | null; active?: boolean; category_ids: string[];
}
export type TenderStatus = "draft" | "sent" | "bids_in" | "compared" | "awarded" | "cancelled";
export type TenderDocumentType = "plan" | "markup" | "spec";
export interface TenderDocument { id: string; tender_package_id: string; document_type: TenderDocumentType; file_path: string; original_filename: string; uploaded_at: string }
export type TenderBidStatus = "invited" | "received" | "reviewed" | "cancelled";
export interface TenderLineItem { description: string; amount: string }
export interface TenderBidDocument { id: string; tender_bid_id: string; file_path: string; original_filename: string; uploaded_at: string }
export interface TenderBid {
  id: string; tender_package_id: string; contractor_id: string; contractor: Contractor; status: TenderBidStatus;
  quote_amount?: string | null; extracted_amount?: string | null; extracted_line_items?: TenderLineItem[] | null;
  excluded_scope_notes?: string | null; reviewer_notes?: string | null; invited_at?: string | null; received_at?: string | null;
  documents: TenderBidDocument[]; created_at: string; updated_at: string;
}
export interface TenderAward {
  id: string; tender_package_id: string; winning_bid_id: string; po_id: string; award_instructions: string;
  project_start_date: string; contractor_start_date: string; awarded_at: string;
  purchase_order?: { id: string; po_number: string; budget_id: string; budget_line_id: string; description: string; amount: string; status: string } | null;
}
export interface TenderPackage {
  id: string; property_id: string; category_id: string; category: ContractorCategory; scope_description: string;
  status: TenderStatus; due_date?: string | null; documents: TenderDocument[]; bids: TenderBid[]; award?: TenderAward | null; created_at: string; updated_at: string;
}
